kolibri/plugins/hooks.py

```diff
@@ -1,59 +1,230 @@
+
+    # This is where the actual abstract hook is defined
+    from kolibri.core.hooks import NavigationHook
+
+    # By inheriting NavigationHook, we tell that we are going to want our
+    # plugin to be part of the hook's activities with the specified attributes.
+    # We only define one navigation item, but defining another one is as simple
+    # as adding another class definition.
+    class MyPluginNavigationItem(NavigationHook):
+
+        label = _("My Plugin")
+        url = reverse_lazy("kolibri:my_plugin:index")
+
+
+And here is the definition of that hook in kolibri.core.hooks:
+
+.. code-block:: python
+
+    from kolibri.plugins.hooks import KolibriHook
+
+
+    class NavigationHook(KolibriHook):
+        \"\"\"
+        Extend this hook to define a new navigation item
+        \"\"\"
+
+        #: A string label for the menu item
+        label = "Untitled"
+
+        #: A string or lazy proxy for the url
+        url = "/"
+
+        @classmethod
+        def get_menu(cls):
+            menu = {}
+            for inheritor in cls.inheritors:
+                menu[inheritor.label] = url
+            return menu
+
+
+        class Meta:
+
+            abstract = True
+
+
+Usage of the hook
+-----------------
+
+Inside our templates, we load a template tag from navigation_tags, and this
+template tag definition looks like this:
+
+.. code-block:: python
+
+    from kolibri.core.hooks import NavigationHook
+
+    @register.assignment_tag()
+    def kolibri_main_navigation():
+
+        for label, url in NavigationHook.get_menu().items():
+            yield label, url
+
+
+.. code-block:: html
+
+    {% load kolibri_tags %}
+
+    <ul>
+    {% for navigation in kolibri_main_navigation %}
+        <li><a href="{{ navigation.menu_url }}">{{ navigation.menu_name }}</a></li>
+    {% endfor %}
+    </ul>
+
+
+.. warning::
+
+    Do not register callbacks for hooks outside of a plugin's
+    ``kolibri_plugin``. Callbacks should all be registered at load time, and
+    placing that logic in ``kolibri_plugin`` guarantees that things are
+    registered correctly.
+
+
 """
 from __future__ import absolute_import, print_function, unicode_literals
 
-__hook_index = -1
+import six
 
 
-def __enumerate_hook():
-    global __hook_index
-    __hook_index += 1
-    return __hook_index
+class KolibriHookMeta(type):
+    """
+    WIP!!!
 
-# : Assign to a callable ``my_func(navigation)``.
-NAVIGATION_POPULATE = __enumerate_hook()
-FRONTEND_PLUGINS = __enumerate_hook()
-BASE_FRONTEND_SYNC = __enumerate_hook()
-BASE_FRONTEND_ASYNC = __enumerate_hook()
+    This is the base meta class of all hooks in Kolibri.
 
-REGISTERED_HOOKS = {
-    x: [] for x in range(__hook_index + 1)
-}
+        Metaclasses are deeper magic than 99% of users should ever worry about.
+        If you wonder whether you need them, you don't (the people who actually
+        need them know with certainty that they need them, and don't need an
+        explanation about why).
```

Comment:
The example should illustrate the method of defining hooks as types. Meta classes add a way to do dynamic stuff at the point of instantiating new types. I'm doing this the docs-driven way, so while I flesh out the docs and see that it's a meaning full approach for plugin/hook authors, it will be implemented as well.

http://kolibri.readthedocs.io/en/hooks-as-classes/dev/plugins.html#kolibri-hooks-api